File: examples/lua/markdown_demo.lua
-- Markdown Component Demo
--
-- Demonstrates the new Markdown rendering capabilities with various themes and content types
-- Shows real-time markdown parsing and rendering with interactive features

local kryon = require("../../../bindings/lua/init")
local markdown = require("../../../bindings/lua/markdown")

-- Demo state
local current_file = "README.md"
local current_theme = "light"
local show_source = false
local word_count = 0
local line_count = 0

-- Sample markdown content for demo
local sample_markdown = [[
# Kryon Markdown Component Demo

Welcome to the **Kryon Markdown Component**! This demonstrates rich text formatting capabilities.

## Features Demonstrated

### Text Formatting
You can use *italic*, **bold**, ***bold italic***, and `code` formatting.

### Lists

#### Unordered Lists
- First item
- Second item
  - Nested item
  - Another nested item
- Third item

#### Ordered Lists
1. First step
2. Second step
3. Third step

### Code Blocks

#### Regular Code Block
```nim
import kryon_dsl, markdown

let app = kryonApp:
  Body:
    Markdown:
      source = "# Hello **World**!"
```

#### Syntax Highlighted Code
```javascript
function renderMarkdown(markdown) {
    const ast = parse(markdown);
    return render(ast);
}
```

### Blockquotes

> This is a blockquote. It can span multiple lines and is useful for highlighting quotes or important information.
>
> You can also have nested blockquotes like this.

### Links and Images

Visit the [Kryon documentation](https://kryon.dev) for more information.

### Tables

| Feature | Status | Priority |
|---------|--------|----------|
| Parsing | ✅ Complete | High |
| Rendering | ✅ Complete | High |
| Themes | ✅ Complete | Medium |
| Images | 🔄 In Progress | Medium |
| Links | ✅ Complete | High |

### Horizontal Rules

---

### Raw HTML

<mark>This is highlighted text</mark>

### Subscript and Superscript

H₂O is water. X² is math.

### Strikethrough

~~This text is crossed out~~

### Emojis

🚀 Rocket ship 🎨 Art palette 📚 Books

## Interactive Features

The Markdown component supports:
- **Real-time parsing** - Content is parsed immediately
- **Theme switching** - Light, dark, and GitHub themes
- **File loading** - Load markdown from files
- **Link handling** - Click handlers for links
- **Statistics** - Word count, line count, reading time

## Performance

The parser is optimized for:
- **Memory efficiency** - Minimal allocations
- **Speed** - Fast parsing of large documents
- **Cross-platform** - Works from MCUs to desktop

Try the controls below to explore different features!
]]

local function updateStatistics(markdown_text)
    -- Count words (split by whitespace)
    local words = {}
    for word in markdown_text:gmatch("%S+") do
        table.insert(words, word)
    end
    word_count = #words

    -- Count lines
    line_count = 0
    for _ in markdown_text:gmatch("\n") do
        line_count = line_count + 1
    end
    line_count = line_count + 1  -- Add 1 for the last line
end

local function createMarkdownFromContent(content, theme_name)
    local theme_obj = nil
    if theme_name == "dark" then
        theme_obj = markdown.themeDark()
    elseif theme_name == "github" then
        theme_obj = markdown.themeGitHub()
    else
        theme_obj = markdown.themeLight()
    end

    return kryon.markdown({
        source = content,
        width = 700,
        height = 600,
        theme = theme_obj,
        on_link_click = function(url)
            print("Link clicked:", url)
        end
    })
end

local function loadMarkdownFile(filename)
    local file = io.open(filename, "r")
    if file then
        local content = file:read("*all")
        file:close()
        updateStatistics(content)
        return content
    else
        return "# File Not Found\n\nThe file '" .. filename .. "' does not exist."
    end
end

-- Initialize statistics
updateStatistics(sample_markdown)

-- Main application setup
local app = kryon.app({
    width = 1000,
    height = 800,
    title = "Markdown Component Demo",

    body = kryon.container({
        width = 1000,
        height = 800,
        background = 0xf8f9fa,

        children = {
            -- Top control panel
            kryon.container({
                width = 1000,
                height = 80,
                background = 0xffffff,
                x = 0,
                y = 0,

                children = {
                    -- Use row layout for controls
                    kryon.row({
                        x = 20,
                        y = 20,
                        gap = 10,

                        children = {
                            -- File selection
                            kryon.button({
                                text = "README.md",
                                on_click = function()
                                    current_file = "README.md"
                                    updateStatistics(loadMarkdownFile(current_file))
                                end
                            }),

                            kryon.button({
                                text = "Sample.md",
                                on_click = function()
                                    current_file = "sample.md"
                                    updateStatistics(sample_markdown)
                                end
                            }),

                            -- Theme selection
                            kryon.button({
                                text = "Light Theme",
                                on_click = function()
                                    current_theme = "light"
                                end
                            }),

                            kryon.button({
                                text = "Dark Theme",
                                on_click = function()
                                    current_theme = "dark"
                                end
                            }),

                            kryon.button({
                                text = "GitHub Theme",
                                on_click = function()
                                    current_theme = "github"
                                end
                            }),

                            -- Statistics display
                            kryon.container({
                                flexGrow = 1,
                                width = 300,
                                height = 40,
                                background = 0xe9ecef,
                                x = 400,
                                y = 20,

                                children = {
                                    kryon.text({
                                        content = "Words: " .. word_count .. " | Lines: " .. line_count .. " | Reading time: " .. math.floor(word_count / 200) .. " min",
                                        color = 0x495057,
                                        font_size = 12
                                    })
                                }
                            }),

                            -- Toggle source view
                            kryon.button({
                                text = show_source and "Hide Source" or "Show Source",
                                on_click = function()
                                    show_source = not show_source
                                end
                            })
                        }
                    })
                }
            }),

            -- Main content area (positioned below control panel)
            kryon.container({
                width = 1000,
                height = 620,
                x = 0,
                y = 80,

                children = {
                    -- Conditional content based on show_source
                    function()
                        if show_source then
                            -- Show source code side by side
                            return {
                                -- Source code panel
                                kryon.container({
                                    width = 500,
                                    height = 600,
                                    background = 0xf1f3f4,
                                    x = 0,
                                    y = 0,

                                    children = {
                                        kryon.text({
                                            content = "Source Code:",
                                            color = 0x495057,
                                            font_size = 16,
                                            x = 20,
                                            y = 20
                                        }),

                                        kryon.text({
                                            content = sample_markdown,
                                            color = 0x212529,
                                            font_size = 12,
                                            x = 20,
                                            y = 50
                                        })
                                    }
                                }),

                                -- Markdown content panel
                                kryon.container({
                                    width = 500,
                                    height = 600,
                                    background = 0xffffff,
                                    x = 500,
                                    y = 0,

                                    children = {
                                        createMarkdownFromContent(
                                            current_file == "sample.md" and sample_markdown or loadMarkdownFile(current_file),
                                            current_theme
                                        )
                                    }
                                })
                            }
                        else
                            -- Show only markdown content
                            return {
                                kryon.container({
                                    width = 1000,
                                    height = 600,
                                    background = 0xffffff,
                                    x = 0,
                                    y = 0,

                                    children = {
                                        createMarkdownFromContent(
                                            current_file == "sample.md" and sample_markdown or loadMarkdownFile(current_file),
                                            current_theme
                                        )
                                    }
                                })
                            }
                        end
                    end
                }
            }),

            -- Footer information
            kryon.container({
                width = 1000,
                height = 60,
                background = 0xe9ecef,
                x = 0,
                y = 700,

                children = {
                    kryon.column({
                        x = 20,
                        y = 20,
                        gap = 5,

                        children = {
                            kryon.text({
                                content = "Markdown Component Demo",
                                color = 0x495057,
                                font_size = 14
                            }),

                            kryon.text({
                                content = "Supports CommonMark + GitHub Flavored Markdown • Real-time parsing • Multiple themes • Interactive features",
                                color = 0x6c757d,
                                font_size = 12
                            })
                        }
                    })
                }
            })
        }
    })
})

print("Markdown Component Demo Started!")
print("Features demonstrated:")
print("- Full CommonMark compliance with GitHub Flavored Markdown extensions")
print("- Real-time parsing and rendering")
print("- Multiple built-in themes (light, dark, GitHub)")
print("- File loading and content display")
print("- Interactive link handling")
print("- Statistics tracking (words, lines, reading time)")
print("- Source code view toggle")
print("- Responsive layout with control panel")
print("")
print("Try the controls above to explore different features!")

File: examples/nim/markdown_demo.nim
## Markdown Component Demo
##
## Demonstrates the new Markdown rendering capabilities with various themes and content types
## Shows real-time markdown parsing and rendering with interactive features

import std/[strutils, tables, os]
import kryon_dsl
import markdown

# Demo state
var
  currentFile = "README.md"
  currentTheme = "light"
  showSource = false
  wordCount = 0
  lineCount = 0

# Sample markdown content for demo
let sampleMarkdown = """
# Kryon Markdown Component Demo

Welcome to the **Kryon Markdown Component**! This demonstrates rich text formatting capabilities.

## Features Demonstrated

### Text Formatting
You can use *italic*, **bold**, ***bold italic***, and `code` formatting.

### Lists

#### Unordered Lists
- First item
- Second item
  - Nested item
  - Another nested item
- Third item

#### Ordered Lists
1. First step
2. Second step
3. Third step

### Code Blocks

#### Regular Code Block
```nim
import kryon_dsl, markdown

let app = kryonApp:
  Body:
    Markdown:
      source = "# Hello **World**!"
```

#### Syntax Highlighted Code
```javascript
function renderMarkdown(markdown) {
    const ast = parse(markdown);
    return render(ast);
}
```

### Blockquotes

> This is a blockquote. It can span multiple lines and is useful for highlighting quotes or important information.
>
> You can also have nested blockquotes like this.

### Links and Images

Visit the [Kryon documentation](https://kryon.dev) for more information.

### Tables

| Feature | Status | Priority |
|---------|--------|----------|
| Parsing | ✅ Complete | High |
| Rendering | ✅ Complete | High |
| Themes | ✅ Complete | Medium |
| Images | 🔄 In Progress | Medium |
| Links | ✅ Complete | High |

### Horizontal Rules

---

### Raw HTML

<mark>This is highlighted text</mark>

### Subscript and Superscript

H₂O is water. X² is math.

### Strikethrough

~~This text is crossed out~~

### Emojis

🚀 Rocket ship 🎨 Art palette 📚 Books

## Interactive Features

The Markdown component supports:
- **Real-time parsing** - Content is parsed immediately
- **Theme switching** - Light, dark, and GitHub themes
- **File loading** - Load markdown from files
- **Link handling** - Click handlers for links
- **Statistics** - Word count, line count, reading time

## Performance

The parser is optimized for:
- **Memory efficiency** - Minimal allocations
- **Speed** - Fast parsing of large documents
- **Cross-platform** - Works from MCUs to desktop

Try the controls below to explore different features!
"""

proc updateStatistics(markdown: string) =
  wordCount = markdown.countWords()
  lineCount = markdown.countLines()

proc createMarkdownFromContent(content: string, theme: string): KryonComponent =
  let themeObj = case theme.toLowerAscii():
    of "dark": darkTheme()
    of "github": githubTheme()
    else: lightTheme()

  Markdown:
    source = content
    theme = themeObj
    width = 700
    height = 600
    onLinkClick = proc(url: string) =
      echo "Link clicked: ", url

proc loadMarkdownFile(filename: string): string =
  if fileExists(filename):
    try:
      result = readFile(filename)
      updateStatistics(result)
    except:
      result = "# Error loading file: " & filename & "\n\nFile may not exist or may not be readable."
  else:
    result = "# File Not Found\n\nThe file '" & filename & "' does not exist."

let app = kryonApp:
  Header:
    windowWidth = 1000
    windowHeight = 800
    windowTitle = "Markdown Component Demo"

  Body:
    backgroundColor = "#f8f9fa"

    # Top control panel
    Container:
      width = 1000
      height = 80
      backgroundColor = "#ffffff"
      padding = 20

      Row:
        # File selection
        Container:
          width = 200
          height = 40

          Button:
            text = "README.md"
            onClick = proc() =
              currentFile = "README.md"
              updateStatistics(loadMarkdownFile(currentFile))

        Button:
          text = "Sample.md"
          onClick = proc() =
              currentFile = "sample.md"
              updateStatistics(sampleMarkdown)

        # Theme selection
        Container:
          width = 200
          height = 40
          marginLeft = 20

          Button:
            text = "Light Theme"
            onClick = proc() = currentTheme = "light"

          Button:
            text = "Dark Theme"
            onClick = proc() = currentTheme = "dark"

          Button:
            text = "GitHub Theme"
            onClick = proc() = currentTheme = "github"

        # Statistics
        Container:
          flexGrow = 1
          height = 40
          backgroundColor = "#e9ecef"
          padding = 10

          Text:
            content = "Words: " & $wordCount & " | Lines: " & $lineCount & " | Reading time: " & $(wordCount div 200) & " min"
            color = "#495057"

        # Toggle source view
        Container:
          width = 120
          height = 40

          Button:
            text = if showSource: "Hide Source" else: "Show Source"
            onClick = proc() = showSource = not showSource

    # Main content area
    Row:
      # Markdown content
      Container:
        width = if showSource: 500 else: 1000
        flexGrow = 1
        backgroundColor = "#ffffff"
        padding = 20

        if showSource:
          # Show source code
          Text:
            content = sampleMarkdown
            color = "#495057"
            fontSize = 12
        else:
          # Show rendered markdown
          createMarkdownFromContent(
            if currentFile == "sample.md": sampleMarkdown
            else: loadMarkdownFile(currentFile),
            currentTheme
          )

      # Source code panel (when shown)
      if showSource:
        Container:
          width = 500
          backgroundColor = "#f1f3f4"
          padding = 20

          Text:
            content = "Source Code:"
            color = "#495057"
            fontSize = 16
            marginBottom = 10

          Text:
            content = sampleMarkdown
            color = "#212529"
            fontSize = 12

    # Footer information
    Container:
      width = 1000
      height = 60
      backgroundColor = "#e9ecef"
      padding = 20

      Column:
        Text:
          content = "Markdown Component Demo"
          color = "#495057"
          fontSize = 14
          alignSelf = "center"

        Text:
          content = "Supports CommonMark + GitHub Flavored Markdown • Real-time parsing • Multiple themes • Interactive features"
          color = "#6c757d"
          fontSize = 12
          alignSelf = "center"

echo "Markdown Component Demo Started!"
echo "Features demonstrated:"
echo "- Full CommonMark compliance with GitHub Flavored Markdown extensions"
echo "- Real-time parsing and rendering"
echo "- Multiple built-in themes (light, dark, GitHub)"
echo "- File loading and content display"
echo "- Interactive link handling"
echo "- Statistics tracking (words, lines, reading time)"
echo "- Source code view toggle"
echo "- Responsive layout with control panel"
echo ""
echo "Try the controls above to explore different features!"